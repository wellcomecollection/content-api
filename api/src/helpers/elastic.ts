import {
  AggregationsAggregate,
  AggregationsStringTermsAggregate,
} from "@elastic/elasticsearch/lib/api/types";

export const isTermsAggregation = (
  agg: AggregationsAggregate
): agg is AggregationsStringTermsAggregate =>
  "buckets" in agg && "doc_count" in agg.buckets[0] && "key" in agg.buckets[0];
