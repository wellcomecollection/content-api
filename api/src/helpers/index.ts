// The built-in types for `set.has` don't include a type guard
export const isInSet =
  <T, S extends T>(set: ReadonlySet<S>) =>
  (item: T): item is S =>
    set.has(item as S);

export const ifDefined = <T, R>(
  maybeParam: T | undefined,
  func: (param: T) => R
): R | undefined => (maybeParam !== undefined ? func(maybeParam) : undefined);

export const isNotUndefined = <T>(val: T | undefined): val is T =>
  val !== undefined;

export const pick = <T extends object, K extends keyof T>(
  obj: T,
  keys: K[]
): Pick<T, K> =>
  Object.fromEntries(
    keys.filter(key => key in obj).map(key => [key, obj[key]])
  ) as Pick<T, K>;

export const not =
  <T extends any[]>( // eslint-disable-line @typescript-eslint/no-explicit-any
    f: (...params: T) => boolean
  ) =>
  (...params: T) =>
    !f(...params);
