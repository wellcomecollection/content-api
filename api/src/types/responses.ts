export type Aggregations = Record<string, Aggregation>;
export type Aggregation = {
  buckets: AggregationBucket[];
  type: "Aggregation";
};
export type AggregationBucket = {
  data: any;
  count: number;
  type: "AggregationBucket";
};

export type ResultList<Result = any> = {
  type: "ResultList";
  results: Result[];
  aggregations?: Aggregations;
  totalResults: number;
  totalPages: number;
  pageSize: number;
};
