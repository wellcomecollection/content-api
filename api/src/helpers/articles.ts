import { ContentType } from "../types";

export const graphQueryArticles = `{
    articles {
      title
      format {
        title
      }
      promo
      contributors {
        ...contributorsFields
        role {
          title
        }
        contributor {
          ... on people {
            name
          }
          ... on organisations {
            name
          }
        }
      }
    }
  }`.replace(/\n(\s+)/g, "\n"); // Pre-emptive removal of whitespaces as requests to the Prismic Rest API are limited to 2048 characters.

export const articlesContentTypes: ContentType[] = ["articles", "webcomics"];
