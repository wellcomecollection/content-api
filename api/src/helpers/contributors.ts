import { TransformedContributor } from "../types";

export const getContributors = (
  contributors: any[]
): TransformedContributor[] => {
  return contributors.map((c): TransformedContributor => {
    return {
      type: "Contributor",
      contributor: {
        id: c.contributor.id, // TODO can we simplify this
        label: c.contributor.data?.name, // TODO can we simplify this
        type: "Person",
      },
      role: {
        id: c.role.id,
        label: c.role.data?.title[0].text, // TODO can we simplify this
        type: "EditorialContributorRole",
      },
    };
  });
};
