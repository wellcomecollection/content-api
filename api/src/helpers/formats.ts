import { FilledLinkToDocumentField } from "@prismicio/types";
import { asText } from ".";
import { ArticleFormat, ArticleFormatId, InferDataInterface, LabelField } from "../types";

export function transformLabelType(
    format: FilledLinkToDocumentField<
      'article-formats',
      'en-gb',
      InferDataInterface<ArticleFormat>
    > & { data: InferDataInterface<ArticleFormat> }
  ): LabelField {
    return {
        type: "ArticleFormat",
        id: format.id as ArticleFormatId,
        title: asText(format.data.title),
    };
  }