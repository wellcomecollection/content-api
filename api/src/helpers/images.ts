import { EmptyImageFieldImage } from "@prismicio/types";
import { PrismicImage } from "../types";

// when images have crops, event if the image isn't attached, we get e.g.
// { '32:15': {}, '16:9': {}, square: {} }
export function isImageLink(
  maybeImage: EmptyImageFieldImage | PrismicImage | undefined
): maybeImage is PrismicImage {
  return Boolean(maybeImage && maybeImage.dimensions);
}
