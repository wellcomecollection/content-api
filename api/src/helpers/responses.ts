import { Request } from "express";
import {
  AggregationsAggregate,
  AggregationsStringTermsBucket,
  SearchResponse,
} from "@elastic/elasticsearch/lib/api/types";
import { Displayable } from "../types";
import {
  AggregationBucket,
  Aggregations,
  ResultList,
} from "../types/responses";
import { Config } from "../../config";
import { paginationResponseGetter } from "../controllers/pagination";
import { isNotUndefined } from "./index";

const mapBucket = (
  bucket: AggregationsStringTermsBucket,
): AggregationBucket => ({
  // This should always be a JSONified string, so if we can't
  // parse it something has gone wrong in the pipeline and we
  // should know about it
  data: JSON.parse(bucket.key),
  count: bucket.doc_count,
  type: "AggregationBucket",
});

// Sort by count (descending) and then by ID (ascending)
const compareBucket = (a: AggregationBucket, b: AggregationBucket) =>
  b.count - a.count || (a.data.id ?? "").localeCompare(b.data.id ?? "");

export const mapAggregations = (
  elasticAggs: AggregationsAggregate,
): Aggregations =>
  Object.fromEntries(
    Object.entries(elasticAggs).flatMap(([name, aggregation]) => {
      const buckets: AggregationsStringTermsBucket[] =
        aggregation.buckets ?? aggregation.terms.buckets;
      const selfFilterBuckets: AggregationsStringTermsBucket[] =
        aggregation.self_filter?.terms.buckets ?? [];

      const bucketKeys = new Set<string>(); // prevent duplicates from the self-filter
      const allBuckets = [...buckets, ...selfFilterBuckets]
        .filter((b) => {
          const result = isNotUndefined(b) && !bucketKeys.has(b.key);
          bucketKeys.add(b?.key);
          return result;
        })
        .map(mapBucket)
        .sort(compareBucket);
      return [[name, { buckets: allBuckets, type: "Aggregation" }]];
    }),
  );

export const resultListResponse = (config: Config) => {
  const getPaginationResponse = paginationResponseGetter(config.publicRootUrl);

  return <R extends Request<any, any, any, any>>(
    req: R,
    searchResponse: SearchResponse<Displayable>,
  ): ResultList => {
    const results = searchResponse.hits.hits.flatMap((hit) =>
      hit._source ? [hit._source.display] : [],
    );

    const aggregations: Aggregations | undefined = searchResponse.aggregations
      ? mapAggregations(searchResponse.aggregations)
      : undefined;

    const requestUrl = new URL(
      req.url,
      `${req.protocol}://${req.headers.host}`,
    );

    const totalResults =
      typeof searchResponse.hits.total === "number"
        ? searchResponse.hits.total
        : (searchResponse.hits.total?.value ?? 0);

    const paginationResponse = getPaginationResponse({
      requestUrl,
      totalResults,
    });

    return {
      type: "ResultList",
      results,
      aggregations,
      ...paginationResponse,
    };
  };
};
