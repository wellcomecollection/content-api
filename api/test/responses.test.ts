import { mapAggregations } from "../src/helpers/responses";

describe("mapAggregations", () => {
  it("maps aggregations responses from ES to our expected response format", () => {
    const elasticAggregations = {
      format: {
        doc_count_error_upper_bound: 0,
        sum_other_doc_count: 0,
        buckets: [
          {
            key: '{"type":"ArticleFormat","id":"W7TfJRAAAJ1D0eLK","label":"Article"}',
            doc_count: 595,
          },
          {
            key: '{"type":"ArticleFormat","id":"W7d_ghAAALWY3Ujc","label":"Comic"}',
            doc_count: 271,
          },
          {
            key: '{"type":"ArticleFormat","id":"W5uKaCQAACkA3C0T","label":"In pictures"}',
            doc_count: 137,
          },
          {
            key: '{"type":"ArticleFormat","id":"W8CbPhEAAB8Nq4aG","label":"Book extract"}',
            doc_count: 38,
          },
          {
            key: '{"type":"ArticleFormat","id":"XTYCkRAAACUANeph","label":"Photo story"}',
            doc_count: 19,
          },
          {
            key: '{"type":"ArticleFormat","id":"XwRZ6hQAAG4K-bbt","label":"Podcast"}',
            doc_count: 16,
          },
          {
            key: '{"type":"ArticleFormat","id":"W9BoHhIAANBp1EXg","label":"Interview"}',
            doc_count: 10,
          },
          {
            key: '{"type":"ArticleFormat","id":"ZBH6PRQAAIrrFirA","label":"Short film"}',
            doc_count: 3,
          },
          {
            key: '{"type":"ArticleFormat","id":"YxcjgREAACAAkjBg","label":"Long read"}',
            doc_count: 2,
          },
          {
            key: '{"type":"ArticleFormat","id":"YrwCTxEAACUALJQ0","label":"Prose poem"}',
            doc_count: 1,
          },
        ],
      },
    };
    const mappedAggregations = mapAggregations(elasticAggregations);
    expect(mappedAggregations.format.buckets).toHaveLength(
      elasticAggregations.format.buckets.length
    );
    expect(mappedAggregations).toMatchSnapshot();
  });

  it("gets buckets from a terms sub-aggregation if the top level aggregation is a single-bucket filter aggregation", () => {
    const elasticAggregations = {
      format: {
        doc_count: 19,
        terms: {
          doc_count_error_upper_bound: 0,
          sum_other_doc_count: 0,
          buckets: [
            {
              key: '{"type":"ArticleFormat","id":"W7TfJRAAAJ1D0eLK","label":"Article"}',
              doc_count: 14,
            },
            {
              key: '{"type":"ArticleFormat","id":"W5uKaCQAACkA3C0T","label":"In pictures"}',
              doc_count: 5,
            },
          ],
        },
      },
    };
    const mappedAggregations = mapAggregations(elasticAggregations);
    expect(mappedAggregations.format.buckets[0].count).toBe(
      elasticAggregations.format.terms.buckets[0].doc_count
    );
    expect(mappedAggregations.format.buckets[1].count).toBe(
      elasticAggregations.format.terms.buckets[1].doc_count
    );
  });

  it("adds self-filter buckets to the list", () => {
    const elasticAggregations = {
      format: {
        doc_count: 7,
        terms: {
          doc_count_error_upper_bound: 0,
          sum_other_doc_count: 0,
          buckets: [
            {
              key: '{"type":"ArticleFormat","id":"W7TfJRAAAJ1D0eLK","label":"Article"}',
              doc_count: 7,
            },
          ],
        },
        self_filter: {
          doc_count: 0,
          terms: {
            doc_count_error_upper_bound: 0,
            sum_other_doc_count: 0,
            buckets: [
              {
                key: '{"type":"ArticleFormat","id":"W7d_ghAAALWY3Ujc","label":"Comic"}',
                doc_count: 1,
              },
              {
                key: '{"type":"ArticleFormat","id":"ZBH6PRQAAIrrFirA","label":"Short film"}',
                doc_count: 0,
              },
            ],
          },
        },
      },
    };

    const mappedAggregations = mapAggregations(elasticAggregations);
    expect(mappedAggregations.format.buckets).toHaveLength(3);
    expect(mappedAggregations.format.buckets).toContainEqual({
      count: 0,
      data: {
        id: "ZBH6PRQAAIrrFirA",
        label: "Short film",
        type: "ArticleFormat",
      },
      type: "AggregationBucket",
    });
    expect(mappedAggregations.format.buckets).toContainEqual({
      count: 1,
      data: {
        id: "W7d_ghAAALWY3Ujc",
        label: "Comic",
        type: "ArticleFormat",
      },
      type: "AggregationBucket",
    });
  });

  it("does not duplicate self-filter buckets", () => {
    const elasticAggregations = {
      format: {
        doc_count: 19,
        self_filters: {
          doc_count: 5,
          format: {
            doc_count_error_upper_bound: 0,
            sum_other_doc_count: 0,
            buckets: [
              {
                key: '{"type":"ArticleFormat","id":"W5uKaCQAACkA3C0T","label":"In pictures"}',
                doc_count: 5,
              },
            ],
          },
        },
        terms: {
          doc_count_error_upper_bound: 0,
          sum_other_doc_count: 0,
          buckets: [
            {
              key: '{"type":"ArticleFormat","id":"W7TfJRAAAJ1D0eLK","label":"Article"}',
              doc_count: 14,
            },
            {
              key: '{"type":"ArticleFormat","id":"W5uKaCQAACkA3C0T","label":"In pictures"}',
              doc_count: 5,
            },
          ],
        },
      },
    };

    const mappedAggregations = mapAggregations(elasticAggregations);
    const bucketKeys = mappedAggregations.format.buckets.map((b) => b.data);
    expect(new Set(bucketKeys).size).toBe(bucketKeys.length);
  });

  it("returns buckets in descending order of count and ascending order of ID", () => {
    const elasticAggregations = {
      "contributors.contributor": {
        doc_count: 37,
        terms: {
          doc_count_error_upper_bound: 0,
          sum_other_doc_count: 20,
          buckets: [
            {
              key: '{"type":"Person","id":"YnjtxxAAACMAHZ_e","label":"Georgie Evans"}',
              doc_count: 6,
            },
            {
              key: '{"type":"Person","id":"Ynz5MBAAAJ-4L-MU","label":"Nicole Coffield"}',
              doc_count: 6,
            },
            {
              key: '{"type":"Person","id":"XFMf5hUAAPGIqj79","label":"Thomas S G Farnetti"}',
              doc_count: 4,
            },
            {
              key: '{"type":"Person","id":"W5e9vyYAACYAMqhi","label":"Elena Carter"}',
              doc_count: 3,
            },
            {
              key: '{"type":"Person","id":"XRShxhEAACMAMrCm","label":"Steven Pocock"}',
              doc_count: 3,
            },
            {
              key: '{"type":"Person","id":"XA5EPBEAAMr9xwoh","label":"Kate Wilkinson"}',
              doc_count: 2,
            },
            {
              key: '{"type":"Person","id":"XBvc8RAAAEpjoYeX","label":"Benjamin Gilbert"}',
              doc_count: 2,
            },
            {
              key: '{"type":"Person","id":"XF1PQhAAAJIpjoT7","label":"Camilla Greenwell"}',
              doc_count: 2,
            },
            {
              key: '{"type":"Person","id":"W-VbKhEAAJWKgZDR","label":"Stevyn Colgan"}',
              doc_count: 1,
            },
            {
              key: '{"type":"Person","id":"W18yKyYAACUAz3oo","label":"Jamie Hale"}',
              doc_count: 1,
            },
            {
              key: '{"type":"Person","id":"W1crUiYAACYArC-y","label":"Sarifa Patel"}',
              doc_count: 1,
            },
            {
              key: '{"type":"Person","id":"W2lv4yYAACQAXk5G","label":"Kristin Hohenadel"}',
              doc_count: 1,
            },
            {
              key: '{"type":"Person","id":"W2q3eCkAACkA7XGA","label":"Lil Sullivan"}',
              doc_count: 1,
            },
            {
              key: '{"type":"Person","id":"W6ooGxIAACMAoWGo","label":"Taras Young"}',
              doc_count: 1,
            },
            {
              key: '{"type":"Person","id":"W8RPPBEAAE8Eu8Gj","label":"Ken Hollings"}',
              doc_count: 1,
            },
            {
              key: '{"type":"Person","id":"WSQ9rygAAA9xtwgQ","label":"Elissavet Ntoulia"}',
              doc_count: 1,
            },
            {
              key: '{"type":"Person","id":"WSRjBCgAAKpwt6w9","label":"Helen Babbs"}',
              doc_count: 1,
            },
            {
              key: '{"type":"Person","id":"WT5y6S0AACwAdxSL","label":"Anna Faherty"}',
              doc_count: 1,
            },
            {
              key: '{"type":"Person","id":"XII-5hAAACrBKerK","label":"Giovanni Tiso"}',
              doc_count: 1,
            },
            {
              key: '{"type":"Person","id":"XS260xAAACUAEfA6","label":"Niven Govinden"}',
              doc_count: 1,
            },
          ],
        },
        self_filter: {
          doc_count: 3,
          terms: {
            doc_count_error_upper_bound: 0,
            sum_other_doc_count: 0,
            buckets: [
              {
                key: '{"type":"Person","id":"XRShxhEAACMAMrCm","label":"Steven Pocock"}',
                doc_count: 3,
              },
            ],
          },
        },
      },
    };

    const mappedAggregations = mapAggregations(elasticAggregations);
    const buckets = mappedAggregations["contributors.contributor"].buckets;
    for (let i = 0; i < buckets.length - 1; i++) {
      expect(buckets[i].count).toBeGreaterThanOrEqual(buckets[i + 1].count);
      if (buckets[i].count === buckets[i + 1].count) {
        expect(
          buckets[i].data.id.localeCompare(buckets[i + 1].data.id)
        ).toBeLessThanOrEqual(0);
      }
    }
  });
});
