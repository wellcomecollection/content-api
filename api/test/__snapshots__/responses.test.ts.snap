// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapAggregations maps aggregations responses from ES to our expected response format 1`] = `
{
  "format": {
    "buckets": [
      {
        "count": 595,
        "data": {
          "id": "W7TfJRAAAJ1D0eLK",
          "label": "Article",
          "type": "ArticleFormat",
        },
        "type": "AggregationBucket",
      },
      {
        "count": 271,
        "data": {
          "id": "W7d_ghAAALWY3Ujc",
          "label": "Comic",
          "type": "ArticleFormat",
        },
        "type": "AggregationBucket",
      },
      {
        "count": 137,
        "data": {
          "id": "W5uKaCQAACkA3C0T",
          "label": "In pictures",
          "type": "ArticleFormat",
        },
        "type": "AggregationBucket",
      },
      {
        "count": 38,
        "data": {
          "id": "W8CbPhEAAB8Nq4aG",
          "label": "Book extract",
          "type": "ArticleFormat",
        },
        "type": "AggregationBucket",
      },
      {
        "count": 19,
        "data": {
          "id": "XTYCkRAAACUANeph",
          "label": "Photo story",
          "type": "ArticleFormat",
        },
        "type": "AggregationBucket",
      },
      {
        "count": 16,
        "data": {
          "id": "XwRZ6hQAAG4K-bbt",
          "label": "Podcast",
          "type": "ArticleFormat",
        },
        "type": "AggregationBucket",
      },
      {
        "count": 10,
        "data": {
          "id": "W9BoHhIAANBp1EXg",
          "label": "Interview",
          "type": "ArticleFormat",
        },
        "type": "AggregationBucket",
      },
      {
        "count": 3,
        "data": {
          "id": "ZBH6PRQAAIrrFirA",
          "label": "Short film",
          "type": "ArticleFormat",
        },
        "type": "AggregationBucket",
      },
      {
        "count": 2,
        "data": {
          "id": "YxcjgREAACAAkjBg",
          "label": "Long read",
          "type": "ArticleFormat",
        },
        "type": "AggregationBucket",
      },
      {
        "count": 1,
        "data": {
          "id": "YrwCTxEAACUALJQ0",
          "label": "Prose poem",
          "type": "ArticleFormat",
        },
        "type": "AggregationBucket",
      },
    ],
    "type": "Aggregation",
  },
}
`;
